CSSFontFace: Fix uncounted lambda captures failures for Safer C++

vitorroriz opened the pull request:
#### 3a96871dc768f7bbdac2dc5837014c6622d006df
<pre>
CSSFontFace: Fix uncounted lambda captures failures for Safer C++
<a href="https://rdar.apple.com/146429578">rdar://146429578</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=289292">https://bugs.webkit.org/show_bug.cgi?id=289292</a>

Reviewed by Chris Dumez.

1. We are marking the callback for iterateClients as NOESCAPE
as they are not stored/called async. We are adapting redefining
the type of accepted callback as NOESCAPE can only decorate
pointer aguments, in this case a function pointer.

Given that, the lambdas used as callback capture &apos;this&apos; but this should be fine
as the callbacks are not called async.

2. angleAtIndex: needs to protected pointer to CSSFontStyleRangeValue when capturing it.

* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::iterateClients):
(WebCore::calculateItalicRange):

Canonical link: <a href="https://commits.webkit.org/291864@main">https://commits.webkit.org/291864@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f1b5f0ec767163d82e12e467c3a12b3b2dd5ca0f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/94229 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13816 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3579 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/99241 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44756 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/96279 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/14116 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/22246 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71880 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/29224 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/97231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10476 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/85079 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/52226 "Found 1 new API test failure: /WPE/TestWebKitWebView:/webkit/WebKitWebView/save (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/10165 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2764 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/44074 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80397 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/101285 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/21281 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15500 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80886 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21533 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/81092 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/80268 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20014 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24817 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/2184 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/14481 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/21265 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/26444 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20952 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/24412 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22693 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/CSSFontFace.cpp:
```diff
@@ -153,7 +153,7 @@ void CSSFontFace::setFamilies(CSSValueList& family)
     mutableProperties().setProperty(CSSPropertyFontFamily, family);
 
-    iterateClients(m_clients, [&](CSSFontFaceClient& client) {
+    iterateClients(m_clients, [this, protectedThis = Ref { *this }](CSSFontFaceClient& client) {
```
The rules forbid capturing `this` fwiw. You should always use `protectedThis` instead.

vitorroriz replied:
I see. I think we should update https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines#when-passing-an-object-to-non-trivial-function-hold-a-smart-pointer-to-this-object-on-the-stack

The right example shows:
```
// Keep `this` alive.
doAsyncWork([this, protectedThis] {
    foo(); // Member function on `this`.
    bar(); // Member function on `this`.
});

```

rniwa replied:
Oh, capturing `this` is fine when there is a `protectedThis`. When we capture `WeakPtr` instead is when we need to avoid capturing `this`.

geoffreygaren commented on Source/WebCore/css/CSSFontFace.cpp:
```diff
@@ -153,8 +153,8 @@ void CSSFontFace::setFamilies(CSSValueList& family)
 
-    iterateClients(m_clients, [&](CSSFontFaceClient& client) {
-        client.fontPropertyChanged(*this, oldFamily.get());
+    iterateClients(m_clients, [&, protectedThis = Ref { *this }](CSSFontFaceClient& client) {
```
In this case a slightly nicer fix is NOESCAPE:

`template<typename T> void iterateClients(WeakHashSet<CSSFontFaceClient>& clients, NOESCAPE T callback)`

geoffreygaren replied:
There's no real need to update all callers to refcount their captures because we only use them locally, so standard enforcement of local variable safety also ensures capture safety.

vitorroriz replied:
Thanks!

vitorroriz replied:
@geoffreygaren it seems NOESCAPE can only be used for pointers. So the callback would have to be or decay to a function pointer, but then we couldn't capture. Should I go back to protected this?

` error: 'noescape' attribute only applies to pointer arguments [-Werror,-Wignored-attributes]`

cdumez replied:
I provided a solution for this one.

cdumez reviewed (approved):
The code change is good but please fix the commit title / message.

cdumez commented on Source/WebCore/css/CSSFontFace.cpp:
```diff
@@ -53,7 +53,7 @@ namespace WebCore {
 DEFINE_ALLOCATOR_WITH_HEAP_IDENTIFIER(CSSFontFace);
 
-template<typename T> void iterateClients(WeakHashSet<CSSFontFaceClient>& clients, T callback)
+template<typename T> void iterateClients(WeakHashSet<CSSFontFaceClient>& clients, NOESCAPE T callback)
```
Just use this to fix the build:
```
static void iterateClients(WeakHashSet<CSSFontFaceClient>& clients, NOESCAPE const Function<void(CSSFontFaceClient&)>& callback)
{
    for (auto& client : copyToVectorOf<Ref<CSSFontFaceClient>>(clients))
        callback(client);
}
```

cdumez reviewed (approved):
LGTM but please wait on the Safer CPP bot.

vitorroriz committed:
CSSFontFace: Fix uncounted lambda captures failures for Safer C++
rdar://146429578
https://bugs.webkit.org/show_bug.cgi?id=289292

Reviewed by Chris Dumez.

1. We are marking the callback for iterateClients as NOESCAPE
as they are not stored/called async. We are adapting redefining
the type of accepted callback as NOESCAPE can only decorate
pointer aguments, in this case a function pointer.

Given that, the lambdas used as callback capture 'this' but this should be fine
as the callbacks are not called async.

2. angleAtIndex: needs to protected pointer to CSSFontStyleRangeValue when capturing it.

* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::iterateClients):
(WebCore::calculateItalicRange):

Canonical link: https://commits.webkit.org/291864@main
